src/main/java/com/uid2/shared/middleware/AuthMiddleware.java

```diff
@@ -64,36 +65,46 @@ public static IAuthorizable getAuthClient(RoutingContext rc) {
         return (IAuthorizable) rc.data().get(API_CLIENT_PROP);
     }
 
-    private IAuthorizable getAuthKey(String key) {
+    public static <U extends IAuthorizable> U getAuthClient(Class<U> type, RoutingContext rc) {
+        return (U) rc.data().get(API_CLIENT_PROP);
+    }
+
+    public static void setAuthClient(RoutingContext rc, IAuthorizable profile) {
+        rc.data().put(API_CLIENT_PROP, profile);
+        if (profile != null) {
+            rc.data().put(API_CONTACT_PROP, profile.getContact());
+        }
+    }
+
+    private IAuthorizable getAuthClientByKey(String key) {
         return this.authKeyStore.get(key);
     }
 
-    public Handler<RoutingContext> handleV1(Handler<RoutingContext> handler, Role... roles) {
+    public <E> Handler<RoutingContext> handleV1(Handler<RoutingContext> handler, E... roles) {
         if (roles == null || roles.length == 0) {
-            return handler;
+            throw new IllegalArgumentException("must specify at least one role");
         }
-        final AuthHandler h = new AuthHandler(handler, Collections.unmodifiableSet(new HashSet<Role>(Arrays.asList(roles))), this.authKeyStore, true);
+        final RoleBasedAuthorizationProvider<E> authorizationProvider = new RoleBasedAuthorizationProvider<>(Collections.unmodifiableSet(new HashSet<E>(Arrays.asList(roles))));
+        final AuthHandler h = new AuthHandler(handler, this.authKeyStore, authorizationProvider, true);
         return h::handle;
     }
 
-    public Handler<RoutingContext> handle(Handler<RoutingContext> handler, Role... roles) {
+    public <E> Handler<RoutingContext> handle(Handler<RoutingContext> handler, E... roles) {
         if (roles == null || roles.length == 0) {
-            return handler;
+            throw new IllegalArgumentException("must specify at least one role");
```

Comment:
Just a reminder: In UID2Operator we have one violation [here](https://github.com/IABTechLab/uid2-operator/blob/master/src/main/java/com/uid2/operator/vertx/UIDOperatorVerticle.java#L158):
`router.get("/v1/token/refresh").handler(auth.handle(this::handleTokenRefreshV1Async));`

Reply:
The [operator PR](https://github.com/IABTechLab/uid2-operator/pull/9) updates this to:

```
router.get("/v1/token/refresh").handler(auth.handleWithOptionalAuth(this::handleTokenRefreshV1Async));
```

Does this look reasonable to you?